pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -90,117 +104,283 @@ func (c *ClusterMemberController) reconcileMembers(ctx context.Context, recorder
 	}
 
-	etcdHost, err := c.getEtcdPeerHostToScale(podToAdd)
+	// Add a learner member if next peer found
+	peerURL, err := c.getEtcdPeerURLToAdd(ctx)
 	if err != nil {
-		return fmt.Errorf("could not get etcd peer host :%w", err)
+		return fmt.Errorf("could not get etcd peerURL to add :%w", err)
 	}
-
-	if machineForHostPendingDeletion, err := c.isMachinePendingDeletionFor(etcdHost); err != nil {
-		return err
-	} else if machineForHostPendingDeletion {
-		return nil
+	if len(peerURL) > 0 {
+		err = c.etcdClient.MemberAddAsLearner(ctx, peerURL)
+		if err != nil {
+			return fmt.Errorf("failed to add learner member :%w", err)
+		}
 	}
 
-	recorder.Eventf("FoundPodToScale", "found pod to add to etcd membership: %v", podToAdd.Name)
-	err = c.etcdClient.MemberAdd(ctx, fmt.Sprintf("https://%s", net.JoinHostPort(etcdHost, "2380")))
-	if err != nil {
-		return fmt.Errorf("could not add member :%w", err)
+	// Attempt to promote a learner member if there is room for more voting members
+	if err := c.ensureEtcdLearnerPromotion(ctx, recorder); err != nil {
+		return fmt.Errorf("failed to promote learner: %w", err)
 	}
+
 	return nil
 }
 
-func (c *ClusterMemberController) isMachinePendingDeletionFor(etcdHost string) (bool, error) {
-	if isFunctional, err := c.machineAPIChecker.IsFunctional(); err != nil {
-		return false, err
-	} else if !isFunctional {
-		// always return false when the machine API is off
-		// otherwise we won't be able to add any member
-		return false, nil
+// getEtcdPeerURLToAdd returns a PeerURL of the next etcd instance to be added as a learner to the cluster. If
+// no eligible member is found return an empty string.
+func (c *ClusterMemberController) getEtcdPeerURLToAdd(ctx context.Context) (string, error) {
+	members, err := c.etcdClient.MemberList(ctx)
+	if err != nil {
+		return "", err
 	}
 
-	masterMachines, err := c.masterMachineLister.List(c.masterMachineSelector)
+	nodes, err := c.masterNodeLister.List(c.masterNodeSelector)
 	if err != nil {
-		return false, err
+		return "", err
 	}
 
-	// one we have functional machine API, get the corresponding machine and check DeletionTimestamp
-	switch machineForEtcdHost, hasMachine := ceohelpers.IndexMachinesByNodeInternalIP(masterMachines)[etcdHost]; {
-	case hasMachine && machineForEtcdHost.DeletionTimestamp != nil:
-		klog.V(2).Infof("member: %v has a machine that is pending deletion: %v", etcdHost, machineForEtcdHost.Name)
-		return true, nil
-	case !hasMachine:
-		return false, fmt.Errorf("unable to find machine for member: %v", etcdHost)
-	default:
-		// we've found a machine and it is not pending deletion
-		return false, nil
+	for _, node := range nodes {
+		if node.DeletionTimestamp != nil {
+			klog.V(2).Infof("Ignoring node (%s) for new member addition as it's pending deletion", node.Name)
+			continue
+		}
+
+		peerURL, err := c.getPeerURLForNode(node)
+		if err != nil {
+			return "", fmt.Errorf("failed to get peerURL for node: %v", err)
```

Comment:
we could collect all the errors and aggregate at the end. We took this approach in the other controllers mainly to be able to make progress in case one machine/node is broken.

Reply:
Agreed, otherwise this could block promotion if we're failing for an unrelated node here.